rewrite diiscord webhook

Kabiso17 opened the pull request:
## 概述

去除DB table型別衝突的問題，重新建立開發環境並進行開發

## 變更內容

- 新增network.py實作取得使用者真實IP
- 新增discord_webhook.py
- 調整human_resources以及supplies並加上webhook

## 測試方式


1. 啟動開發伺服器
2. 前往 `/docs` 測試新的 API endpoint
3. 驗證回應格式是否正確

## 相關 Issue

https://www.notion.so/283f457cb428803a8828ca6dffd54eaf?v=287f457cb42880d7bd16000c046dc966&p=28bf457cb4288089841fc2b8b3fabd12&pm=s


## 截圖

<img width="1366" height="352" alt="image" src="https://github.com/user-attachments/assets/bab4afe2-8cca-4d85-90d2-6ce3644b570d" />
<img width="782" height="192" alt="image" src="https://github.com/user-attachments/assets/dca8da30-040e-461c-ab92-32584994b484" />

<img width="1407" height="224" alt="image" src="https://github.com/user-attachments/assets/fbceee1d-fd23-44d7-b495-689ec576b93d" />
<img width="606" height="110" alt="image" src="https://github.com/user-attachments/assets/d44a1e2a-2588-4b86-8b02-b93324eab945" />

<img width="1392" height="290" alt="image" src="https://github.com/user-attachments/assets/530c91b8-a704-47a8-b516-292ca59a7b33" />
<img width="485" height="205" alt="image" src="https://github.com/user-attachments/assets/b76194f4-8242-4cfa-b5a2-4dd6c52e9520" />

<img width="1413" height="193" alt="image" src="https://github.com/user-attachments/assets/2f2d9470-257e-4848-b69d-e0dc4708185a" />
<img width="473" height="188" alt="image" src="https://github.com/user-attachments/assets/bcc037b8-5a1f-4ba0-a125-e372797e1c77" />


## Checklist

- [ ] 程式碼遵循專案規範
- [ ] 已新增/更新必要的測試
- [ ] 已更新相關文件
- [ ] CI 檢查全部通過
- [ ] 已在本地測試過
- [ ] Commit 訊息符合 Conventional Commits 規範

Commit:
rewrite diiscord webhook

zenuie commented on guanfu_backend/src/routers/supplies.py:
```diff
@@ -88,7 +104,7 @@ async def create_supply(supply_in: schemas.SupplyCreate, db: Session = Depends(g
     # dependencies=[Security(require_modify_api_key)],
 )
-def patch_supply(id: str, supply_in: schemas.SupplyPatch, db: Session = Depends(get_db)):
+async def patch_supply(id: str, supply_in: schemas.SupplyPatch, request: Request, db: Session = Depends(get_db)):
```
如同copilot所說，在discord_webhook裡面已經有function是用來做format，應該不用在這邊重複寫，直接引用format_supply_patch_notification

Kabiso17 replied:
已更正

zenuie commented on guanfu_backend/src/routers/supplies.py:
```diff
@@ -63,19 +64,34 @@ def list_supplies(
     "", response_model=schemas.SupplyWithPin, status_code=201, summary="建立供應單"
 )
-async def create_supply(supply_in: schemas.SupplyCreate, db: Session = Depends(get_db)):
+async def create_supply(request: Request, supply_in: schemas.SupplyCreate, db: Session = Depends(get_db)):
```
如同copilot所說，在discord_webhook裡面已經有function是用來做format，應該不用在這邊重複寫，直接引用format_supply_notification

Kabiso17 replied:
已更正

Commit:
fix copilot review suggest

Kabiso17 commented on guanfu_backend/src/services/discord_webhook.py:
```diff
@@ -26,7 +221,11 @@ async def send_discord_message(content: str, embed_data: dict | None = None):
     async with httpx.AsyncClient() as client:
         try:
-            await client.post(settings.DISCORD_WEBHOOK_URL, json=message)
+            print(f"Sending Discord webhook: {message}")
```
<img width="475" height="179" alt="image" src="https://github.com/user-attachments/assets/af41bc6a-608f-4aa5-bf5d-a1e8c5aa30a5" />

Kabiso17 commented on guanfu_backend/src/services/discord_webhook.py:
```diff
@@ -26,7 +221,11 @@ async def send_discord_message(content: str, embed_data: dict | None = None):
+            print(f"Sending Discord webhook: {message}")
+            response = await client.post(settings.DISCORD_WEBHOOK_URL, json=message)
+            response.raise_for_status() # Raise an exception for bad status codes
+            print(f"Discord webhook sent successfully. Status: {response.status_code}")
```
<img width="945" height="218" alt="image" src="https://github.com/user-attachments/assets/5ea3d166-83cc-4b31-a52a-f22a55745789" />

Kabiso17 commented on guanfu_backend/src/services/discord_webhook.py:
```diff
@@ -26,7 +221,11 @@ async def send_discord_message(content: str, embed_data: dict | None = None):
+            print(f"Discord webhook sent successfully. Status: {response.status_code}")
         except httpx.RequestError as e:
-            # In a real app, you'd want to log this error.
             print(f"Error sending Discord webhook: {e}")
```
<img width="945" height="218" alt="image" src="https://github.com/user-attachments/assets/5c879fa8-23a3-4eb3-8c90-b6605353d0d4" />

Kabiso17 commented on guanfu_backend/src/services/discord_webhook.py:
```diff
@@ -26,7 +221,11 @@ async def send_discord_message(content: str, embed_data: dict | None = None):
-            # In a real app, you'd want to log this error.
             print(f"Error sending Discord webhook: {e}")
+        except httpx.HTTPStatusError as e:
+            print(f"Error response from Discord webhook: {e.response.status_code} - {e.response.text}")
```
<img width="945" height="218" alt="image" src="https://github.com/user-attachments/assets/a20932d6-93d9-44ca-bec8-866820fc01d1" />

Kabiso17 commented on guanfu_backend/src/routers/supplies.py:
```diff
@@ -98,11 +114,35 @@ def patch_supply(id: str, supply_in: schemas.SupplyPatch, db: Session = Depends(
+        f"IP: {ip_address} (TW)\n"
+        f"User-Agent: {user_agent}"
+    )
+    print(f"Sending Discord webhook from supplies PATCH: {message}")
```
<img width="564" height="342" alt="image" src="https://github.com/user-attachments/assets/421eb259-66be-4479-927b-0799faa3cffc" />

Kabiso17 replied:
<img width="613" height="446" alt="image" src="https://github.com/user-attachments/assets/359ef5e0-4b07-4478-852d-0371aff2c394" />

Commit:
fix some copilot suggest

Commit:
Fix typo

ZhuMon commented on guanfu_backend/src/services/discord_webhook.py:
```diff
@@ -1,15 +1,213 @@
+更新欄位:
+{fields_str}
+IP: {client_ip}
+User-Agent: {user_agent}"""
```
通常 patch 會用在有人提供物資時，而不是改單

Kabiso17 replied:
已更改

WilliamTai commented on guanfu_backend/src/services/discord_webhook.py:
```diff
@@ -26,7 +224,11 @@ async def send_discord_message(content: str, embed_data: dict | None = None):
     async with httpx.AsyncClient() as client:
         try:
-            await client.post(settings.DISCORD_WEBHOOK_URL, json=message)
+            logger.debug(f"Sending Discord webhook: {message}")
```
我們的prod環境會輸出debug level的log嗎？如果沒有，這一行留在這裡只有當作local開發debug用途？

如果我們要在prod環境做troubleshooting，這裡改成info會比較安全。

Kabiso17 replied:
fixed

WilliamTai commented on guanfu_backend/src/services/discord_webhook.py:
```diff
@@ -1,15 +1,213 @@
+
+    Note:
+        如果 DISCORD_WEBHOOK_URL 未設定，函數將直接返回不執行任何操作
     """
     if not settings.DISCORD_WEBHOOK_URL:
         return
```
上層有做輸出Error表示漏掉webhook url設定的訊息嗎？如果沒有，在這裡輸出一個Error/warning message提醒使用者環境變數裡面沒有帶入webhook url，這樣在prod環境debug的時候比較快。

Kabiso17 replied:
fixed

Commit:
fix typo

Commit:
fix: enhance webhook logging and security practices

- Add webhook URL configuration warning
- Standardize headers and logging format
- Improve code consistency

Commit:
remove debug message

ZhuMon commented on guanfu_backend/src/services/discord_webhook.py:
```diff
@@ -1,17 +1,216 @@
+    """
+    # 構建更新欄位列表
+    updates = []
+    if updated_fields.name is not None:
+        updates.append(f"  - name: {updated_fields.name}")
+    if updated_fields.address is not None:
+        updates.append(f"  - address: {updated_fields.address}")
+    if updated_fields.phone is not None:
+        updates.append(f"  - phone: {updated_fields.phone}")
+    if updated_fields.notes is not None:
+        updates.append(f"  - notes: {updated_fields.notes}")
```
如果有人提供了一把鏟子，是不是沒辦法從這個 discord 通知看到

Kabiso17 replied:
已調整程式
新增需求單
<img width="453" height="200" alt="image" src="https://github.com/user-attachments/assets/2b9f7034-27db-4669-ac54-68091e32940b" />
當需求單被變更時
<img width="636" height="289" alt="image" src="https://github.com/user-attachments/assets/d376b7fc-a6e7-462f-a09c-1dd610dcf2be" />

ZhuMon reviewed (commented):
麻煩想一下使用情境喔~

Kabiso17 committed:
fix: enhance supply update notifications and logging

Kabiso17 committed:
fix: update Discord notification handling to use BackgroundTasks

Kabiso17 committed:
fix: reorder parameters in create and patch functions for consistency; enhance role type display formatting in Discord notifications

Kabiso17 committed:
fix error message info
